[RFC] man.vim: bold and underline backspaced text

nhooyr opened the pull request:
\cc @tweekmonster @mhinz 

<img width="1440" alt="screen shot 2016-12-30 at 4 49 01 pm" src="https://cloud.githubusercontent.com/assets/10180857/21573256/4d1e0d16-ceb0-11e6-8019-b50e25657ffd.png">

Fixes #5846 

Conflicts with #5831 but it will be easy to resolve.

nhooyr committed:
man.vim: bold and underline backspaced text

Closes #5846

nhooyr commented on runtime/syntax/man.vim:
```diff
@@ -34,6 +34,10 @@ if b:man_sect =~# '^[23]'
   highlight default link manCFuncDefinition Function
 endif
 
+highlight default manBold cterm=bold gui=bold
```
Should these instead be `manBolded` and `manUnderlined`?

tweekmonster replied:
I think they're better as nouns and not verbs.

mhinz commented on runtime/autoload/man.vim:
```diff
@@ -309,9 +307,39 @@ function! s:format_candidate(path, psect) abort
 endfunction
 
+function! s:strip_backspaced_text(match) abort
+  let s:stripped = substitute(a:match, '.\b', '', 'g')
```
`return substitute()`. :>

mhinz commented:
Parts of the table are highlighted as `manOptionDesc`:

![](https://i.imgur.com/A4tVHyq.png)

nhooyr commented:
@mhinz there are all sorts of small inconsistencies like that. manpage syntax varies a lot, I tested those syntax groups a lot before with many different manpages, I tried to pick the most general rules that would work fine without causing too many problems. The options group in particular was very diverse. I'm not sure how to fix the table while keeping the regexp fairly simple and not causing problems elsewhere. For example, if we say only allow two dashes or plusses, then any characters that are not whitespace and not a dash or a plus, we'll run into a conflict with some options that have a dash in their name. We could disallow a plus followed by a dash but I feel like that is too much of a special case and there is not much benefit. There are probably way more special cases and trying to handle them would very quickly become complicated. I'm gonna leave it alone for now.

mhinz commented:
Yea, if it's not straight-forward to fix, I'd leave it as-is as well.

nhooyr commented:
Upon changing the colorscheme, the highlighting disappears. I'm gonna try using syntax groups instead.

nhooyr commented:
I made one that works on syntax groups but it overrides the highlight of other areas it matches, such as the section titles and options. I'm not sure how to fix this. See https://github.com/nhooyr/neovim/commit/1df9f0defdb3021126ad23b5f0ad14978b5cef68

nhooyr commented:
The highlight group also get cleared by changing the colorscheme with the syntax regions version so it doesn't really fix the original problem and just comes up with more. I'm just not quite sure why the highlight groups are being cleared. hmm

nhooyr committed:
man.vim: simplified man#highlight_backspaced_text

nhooyr committed:
man.vim: handle :colorscheme

nhooyr commented:
Ah, I think this is expected actually, `:highlight clear` should clear the `manBold` and `manUnderline` highlighting groups. We just need to listen to the `ColorScheme` event.

So, it's good now. The syntax region version still has the same problem so I think using the neovim specific bufhl functions is fine unless someone can figure out how to fix the syntax region issue. If we can make that work, I'd rather use that.

nhooyr commented:
Using conceal does have one advantage http://unix.stackexchange.com/a/271905

Is it worth it? Should I switch?

justinmk commented on runtime/autoload/man.vim:
```diff
@@ -309,17 +307,54 @@ function! s:format_candidate(path, psect) abort
+      let pattern = '\%(\(.\)\b\1\)\+'
+      let group = 'manBold'
+    end
+    execute 'silent keepjumps substitute/'.pattern.'/\=s:strip_backspaced_text(submatch(0))'
```
`keeppattern` is also needed.

justinmk commented on runtime/autoload/man.vim:
```diff
@@ -309,17 +307,54 @@ function! s:format_candidate(path, psect) abort
+function! man#highlight_backspaced_text() abort
+  call nvim_buf_clear_highlight(0, s:src_id, 0, -1)
+  while 1
+    let pos = searchpos('\%(_\b[^_]\)\|\%(\(.\)\b\1\)', 'p')
```
`searchpos()` has a "timeout" parameter, but it's not very useful in a while-loop. And calling `reltime()` in a loop will be too expensive. So maybe the loop should be limited to 1000 iterations.

justinmk commented:
Conceal is probably the "right" way to do it. But performance is far more important for man.vim. 

Also adding ~100 lines of code for this feature seems questionable to me. It's important for core plugins to not regress, and we don't have any tests for man.vim, so changes to it are risky.

jamessan commented:
> Conceal is probably the "right" way to do it.

Conceal doesn't hide the characters from the cursor, so movement is awkward.

nhooyr committed:
man.vim: use conceal to hide backspaced text

nhooyr commented:
@justinmk conceal isn't slow, using syntax regions to highlight text is very slow. Before the commit I just pushed, I was using `nvim_buf_add_highlight()` to set the highlight for backspaced text. In another branch on my fork, I was using a syntax region with each start/end pointing to the start/end of some backspaced word. So the loop would keep adding more syntax regions. This was very slow. 

My new approach is based on using conceal and syntax matches and is *very* fast but it has a few different problems. First, movement is awkward with hjkl because of conceal. Second, movement is also awkward with w because neovim scrolls horizontally based on concealed text and we have to set nowrap because neovim doesn't wrap based on visible text, it includes concealed text. See vim/vim#260.

Third, since syntax groups do not combine gracefully, so if a subheading is backspaced to make it bold, only manBold or manSubHeading can highlight it. Since all manSubHeadings are backspaced to make them bold anyway, I could make manSubHeading bold by default but it should still have the color of some default group but since I cannot link it to another group but also make it bold I'd have to use a hacky workaround for that. What I realized is that I can completely eliminate all the syntax regexes since they are redundant with the backspaced highlighting. This is more clean and makes more sense with manpages as the backspaces are there for the proper highlighting. All edge cases with syntax highlighting are eliminated just like that. Though it is definitely less colorful.

With some hackery, I made manBold link to Keyword by default but with `cterm=bold gui=bold`. This isn't a true link since if you change Keyword, manBold will not change automatically to stay in sync but I think most people would prefer color to manBold "linking" to nothing and being just bold.

So, do you guys think these drawbacks worth the performance and "correctness" gains?

Also, would it be easier to fix the conceal issues in neovim compared to vim (make movement and wrapping completely ignore concealed text)?

justinmk commented:
> What I realized is that I can completely eliminate all the syntax regexes since they are redundant with the backspaced highlighting. This is more clean and makes more sense with manpages as the backspaces are there for the proper highlighting. All edge cases with syntax highlighting are eliminated

Nice!

> So, do you guys think these drawbacks worth the performance and "correctness" gains?

What drawbacks? Faster and more elegant sounds good to me :)

> Also, would it be easier to fix the conceal issues in neovim compared to vim (make movement and wrapping completely ignore concealed text)?

Unlikely. Every editor struggles with efficient calculation of text attributes, just like the DOM struggles with node changes, because each change may have cascading effects on neighbors.

justinmk commented:
Getting this error after trying this locally:

```
:Man man             Error detected while processing function <SNR>90_init_highlight_groups:                                 
line   10:                                                                                                                   
E418: Illegal value: boldctermfg=197gui=boldguifg=#F92672
```

justinmk commented on runtime/syntax/man.vim:
```diff
@@ -5,18 +5,30 @@ if exists('b:current_syntax')
+        let values = substitute(values, '.* \(\w\+=.*\)', '\1', '')
+        break
+      endif
+      let group = substitute(values, '.* to \(\w\+\)', '\1', '')
```
This is tricky to get right and I kinda winced when I saw it :) Maybe this helps though: https://github.com/justinmk/vim-sneak/blob/master/autoload/sneak/hl.vim#L19

nhooyr replied:
Instead of this hacky stuff, should I just make manBold link to Keyword and not bold it by default? If a user wants to bold it, they just have to define manBold themselves. Would be much easier.

nhooyr replied:
Or just have it bold by default?

nhooyr replied:
Yea, linking to Keyword by default without bolding doesn't look too bad. 

https://dropbox.aubble.com/Screen%20Shot%202017-01-04%20at%202.58.53%20AM.png

nhooyr replied:
Maybe manUnderline should just link to something by default too instead of underlining it.

nhooyr commented:
@justinmk biggest drawback is less highlight groups for customization. Now, you only get to choose two highlight groups, one for bolded text, one for underlined. Whereas before, you got to choose colors for the title/subheadings/references and what not. Also, the conceal issues.

justinmk commented:
I'd say give the hack a try and if it continues to cause problems that we don't catch here, revert to the simpler approach. 

Underlining and bolding out of the box is nice.

nhooyr committed:
man.vim: bold and underline highlighting cleanup

More robust now.

nhooyr commented:
Done.

justinmk commented on runtime/autoload/man.vim:
```diff
@@ -310,20 +308,37 @@ function! s:format_candidate(path, psect) abort
+endfunction
+
+function! s:get_hl(group) abort
+  redir => hl | execute 'silent highlight' a:group | redir END
```
can use `execute()` here. Crusty old sneak.vim doesn't because it supports Vim 7.2 :)

nhooyr replied:
haha ok

justinmk commented on runtime/syntax/man.vim:
```diff
@@ -5,18 +5,21 @@ if exists('b:current_syntax')
-syntax match manOptionDesc     display '^\s\+\%(+\|-\)\S\+'
+" Not contained because sometimes, manpages contain garbage backspaced
+" characters. See Restricted Shell at bash(1)
+syntax match manBackspacedCharacter display conceal '.\b'
```
`manBackspacedChar` is a bit less wordy.

nhooyr commented:
Done.

tweekmonster commented on runtime/syntax/man.vim:
```diff
@@ -5,18 +5,21 @@ if exists('b:current_syntax')
-syntax match manSubHeading     display '^ \{3\}\S.*$'
-syntax match manOptionDesc     display '^\s\+\%(+\|-\)\S\+'
+" Not contained because sometimes, manpages contain garbage backspaced
+" characters. See Restricted Shell at bash(1)
```
Is this referring to the `+^Ho`?  It's actually not garbage!  I don't remember where I mentioned this, but the backspaces are for moving a print head back to stamp a character over the previous one.  `+^Ho` is a bullet point:

![dots](https://cloud.githubusercontent.com/assets/111942/21754706/765f6baa-d5d3-11e6-90a5-a5e0d00858e3.png)

There's a reference for this somewhere, but I can't remember the URL.

nhooyr replied:
You mentioned it in the original rewrite PR, but I thought it only applied for characters that repeated themselves after the backspace. so `x\bx` where x was a character because this is the regex you used on the PR you made on my fork. But nice catch, I've fixed it now for this PR, those should be highlighted correctly now.

tweekmonster replied:
I guess since this kind of physical printing isn't common any more, it just looks like an inefficient binary instruction.  You might be too young and missed out on watching a dot matrix printer buzzing through its tape on the top shelf of a desk [such as this one][1].  Your heart raced every time your desk shook as the printer fed a new line, loosening, ever so slightly, the screws that held it all together.  It was a rush knowing you were risking your life printing homework because you couldn't find the allen wrench for this out-of-place piece of furniture that was holding a 30lb printer up above your head.  But, oh man was it worth it!  Once it was all done, you tore the curtain of paper away from the printer and gently peeled the trim off the sides.  You were left with a great sense of accomplishment and a handful of trim that your sister used to make bracelets.  You told your sister that her bracelets sucked, but you were secretly jealous of her creativity.  So many emotions swirled around this magic machine.

Anyways, what was I talking about?  Oh yeah, this is called `overstrike` printing.  Dot matrix printers didn't do this.  It was impact/hammer printers like the [daisy wheel][2] or a standard typewriter that used this method.  The print position is backed up by one character and stamped over.  To black out mistakes on a typewriter, X's or #'s were typically used.

I still can't find the reference, but you shouldn't really need it to interpret what you're seeing.  Off the top of my head, these could be created by using characters available on a daisy wheel:

`a^H~` = `ã`
``e^H` `` = `è`
`Y^H=` = `¥`
`C^H=` = `€`
`/^H=` = `≠`
`O^H|` = `ϕ`
`!^H?` = `‽`

Those are just examples to drive the point home for posterity.  I doubt any of those would be found in man pages.

[1]: https://share.esdf.io/HC35Ydk0jY/Songmics-Computer-Desk-Study-Table-Portable-Trolley-Home-_1.jpg
[2]: https://en.wikipedia.org/wiki/Daisy_wheel_printing

nhooyr replied:
Very interesting. And yea, I missed out on all that.

I just bolded any backspaced character that does not have an underline before the backspace since as you said, it's unlikely for these characters to be in man pages and it's simple.

nhooyr committed:
man.vim: more loose syntax matches

All characters that are after a backspace where the character before is
not an underline should be bolded.

nhooyr committed:
man.vim: add backspace to 'iskeyword'

This makes `*` work on headings in man buffers.

nhooyr commented:
Another downside is that you can't search for headings anymore because of the backspaces. That is really unfortunate, I'm not sure if we can get around that.

justinmk commented:
> you can't search for headings anymore because of the backspaces.

Why not delete the backspace characters after using them to calculate the text that needs to be decorated? 

Setting 'iskeyword' is a controversial thing, though `\b` is probably not going to bother anyone.

nhooyr commented:
@justinmk since we're using syntax matches now, we need the backspace characters there so that the regexes match. Can't remove them as far as I can tell.

justinmk commented on runtime/autoload/man.vim:
```diff
@@ -310,20 +308,33 @@ function! s:format_candidate(path, psect) abort
+    endif
+    let group = matchstr(values, '\w\+$')
+  endwhile
+  execute 'highlight default manBold' values 'cterm=bold gui=bold'
```
this is causing an error:

```
:Man ls                                                         Error detected while processing function man#init_highlight_g
roups:                                                                                                                       
line   10:                                                                                                                   
E418: Illegal value: boldctermfg=197gui=boldguifg=#F92672
```

It looks like spaces are getting remove somehow.

jamessan replied:
> It looks like spaces are getting remove somehow.

Yes, that's a regression between Vim and Neovim's implementation of execute.

Vim:

```
:echo execute('hi Normal')
Normal         xxx ctermfg=255 ctermbg=235 guifg=#eeeeee guibg=#262626
```

Neovim:

```
:echo execute('hi Normal')
Normalxxx ctermfg=255ctermbg=235guifg=#eeeeeeguibg=#262626
```

It's related to how the highlighting is handled.  The highlighting obviously can't be represented in the resulting string, but for some reason we're also stripping whitespace in that handling.

justinmk replied:
https://github.com/neovim/neovim/issues/5422

justinmk replied:
@nhooyr This bug was fixed, so rebasing this should work better.

justinmk commented:
Also note that the `execute()` bug will be fixed by https://github.com/neovim/neovim/pull/5975

justinmk commented:
see https://github.com/neovim/neovim/pull/7623
